### PR TITLE
Redis distributed/output cache: include scenario in connection metadata

### PR DESCRIPTION
**Pull request by mgravell**
# Redis distributed/output cache: include scenario in connection metadata and centrailize exception handling

- advertise scenario (output-cache / distributed cache, etc) in the `CLIENT SETINFO` metadata (this is technically a redis 7.2 feature, but a: it is fine to issue on earlier servers - the library already does, and b: Azure Cache for Redis has backported this feature to earlier editions)
- allow initially broken connections (set `AbortOnConnectFail` to `false`)
- centralize output-cache exception handling (much more important now that provider can be out-of-proc); cache being unavailable should be logged but should not interrupt service
- rev se.redis library version (this API is new)
- tweak SignalR API usage because of se.redis lib rev; implicit `string` to `RedisChannel` operator is now `[Obsolete(...)]`, so clarify that we mean single-name channels

The logic for the changes here is moved to a new `internal RedisCacheOptions.GetConfiguredOptions(string libSuffix)` method, for DRY reasons.

### CODE REVIEW

File: src/Middleware/OutputCaching/src/OutputCacheMiddleware.cs [hunk 1]
```diff
@@ -336,10 +336,23 @@ internal async Task<bool> TryServeFromCacheAsync(OutputCacheContext cacheContext
...
+            cacheEntry = await _outputCacheEntryDispatcher.ScheduleAsync(cacheContext.CacheKey, (Store: _store, CacheContext: cacheContext), static async (key, state) => await OutputCacheEntryFormatter.GetAsync(key, state.Store, state.CacheContext.HttpContext.RequestAborted));
+        }
+        catch (OperationCanceledException)
+        {
+            // don't report as failure
+            cacheEntry = null;
+        }
+        catch (Exception ex)
+        {
+            _logger.LogError(ex, "Unable to query output-cache");
```
Comment (BrennanConroy) [inline src/Middleware/OutputCaching/src/OutputCacheMiddleware.cs:351]: Use the same pattern as other logs, i.e. `[LoggerMessage(...)]`.

And end log in `.`

File: src/Middleware/OutputCaching/src/OutputCacheEntryFormatter.cs [hunk 2]
```diff
@@ -85,14 +86,25 @@ public static async ValueTask StoreAsync(string key, OutputCacheEntry value, Tim
...
+        catch (OperationCanceledException)
+        {
+            // don't report as failure
         }
-        else
+        catch (Exception ex)
         {
-            // legacy API/in-proc: create an isolated right-sized byte[] for the payload
-            await store.SetAsync(key, payload.ToArray(), value.Tags, duration, cancellationToken);
+            logger.LogError(ex, "Unable to write to output-cache");
```
Comment (BrennanConroy) [inline src/Middleware/OutputCaching/src/OutputCacheEntryFormatter.cs:107]: Same here

File: src/Middleware/OutputCaching/test/OutputCacheEntryFormatterTests.cs [hunk 3]
```diff
@@ -2,7 +2,7 @@
 // The .NET Foundation licenses this file to you under the MIT license.
```
Comment (BrennanConroy) [inline src/Middleware/OutputCaching/test/OutputCacheEntryFormatterTests.cs:2]: Add OCE and other exception tests?

**Review by BrennanConroy — COMMENTED** [review]
> advertise scenario (output-cache / distributed cache, etc) in the CLIENT SETINFO metadata

Should we add lib-name to SignalR as well?

**Comment by mgravell** [comment]
> > advertise scenario (output-cache / distributed cache, etc) in the CLIENT SETINFO metadata
> 
> Should we add lib-name to SignalR as well?

Done https://github.com/dotnet/aspnetcore/pull/49291/commits/5b40fef3fd4975f56a2d27e28aade638ae31a9bf

File: src/Middleware/OutputCaching/src/LoggerExtensions.cs [hunk 4]
```diff
@@ -49,4 +49,10 @@ internal static partial class LoggerExtensions
         EventName = "ExpirationExpiresExceeded")]
     internal static partial void ExpirationExpiresExceeded(this ILogger logger, DateTimeOffset responseTime);
 
+    [LoggerMessage(12, LogLevel.Error, "Unable to query output cache.", EventName = nameof(UnableToQueryOutputCache))]
```
Comment (BrennanConroy) [inline src/Middleware/OutputCaching/src/LoggerExtensions.cs:52]: Don't use `nameof` for the `EventName`. We treat `EventName` sort of like public API and using `nameof` means the `EventName` will change if we change the function name.

**Review by BrennanConroy — APPROVED** [review]

File: src/Middleware/OutputCaching/src/LoggerExtensions.cs (hunk 4 above)
Comment (mgravell) [reply src/Middleware/OutputCaching/src/LoggerExtensions.cs]: noted; [changing to hard code](https://github.com/dotnet/aspnetcore/pull/49291/commits/96bd5e6da3b26bab0c84ca0673b2b71623bb8870)

**Pull request merged**